File: forense/web/clinical-eye.js
/** Ojo clínico — evaluación del instante en el reproductor forense. */

function trackKindMap(frameRec) {
  const out = new Map();
  for (const tr of frameRec?.tracks || []) {
    if (tr.track_id != null) out.set(tr.track_id, tr.kind || "other");
  }
  return out;
}

function eventsNear(timeline, timeSec, tolerance = 1.5) {
  if (!timeline?.length) return [];
  return timeline.filter((ev) => Math.abs((ev.time_sec || 0) - timeSec) <= tolerance);
}

export function evaluateInstantAudit(frameRec, context = {}) {
  const {
    timeline = [],
    knowledgeMatches = [],
    videoCaption = null,
    timeSec = frameRec?.time_sec,
    minDistanceM = 2,
    maxMachineryKmh = 15,
    maxPersonKmh = 8,
    timeTolerance = 1.5,
  } = context;

  if (!frameRec) {
    return {
      level: "idle",
      statusLabel: "Sin datos",
      headline: "Esperando análisis del video…",
      glanceMetric: "",
      sections: [],
      eventsNear: [],
    };
  }

  const t = timeSec ?? frameRec.time_sec ?? 0;
  const counts = frameRec.counts || {};
  const persons = counts.persons || 0;
  const vehicles = counts.vehicles || 0;
  const prox = frameRec.proximity || [];
  const speeds = frameRec.speeds || [];
  const kindByTrack = trackKindMap(frameRec);
  const near = eventsNear(timeline, t, timeTolerance);

  let level = "ok";
  const issues = [];

  const proxAlert = prox.find((p) => p.alert);
  if (proxAlert) {
    level = "alert";
    issues.push(
      `Proximidad crítica ${proxAlert.distance_m} m (persona #${proxAlert.person_track} – máq. #${proxAlert.machinery_track})`,
    );
  } else if (prox.length) {
    const dist = prox[0].distance_m ?? 99;
    if (dist <= minDistanceM * 1.5) {
      level = "warn";
      issues.push(`Distancia persona–maquinaria: ${dist} m (umbral ${minDistanceM} m)`);
    }
  }

  const speedLines = [];
  for (const sp of speeds) {
    const tid = sp.track_id;
    const kmh = sp.speed_kmh ?? 0;
    const kind = kindByTrack.get(tid) || "other";
    speedLines.push({ tid, kmh, kind, line: `#${tid} ${kmh} km/h` });
    const limit = kind === "machinery" ? maxMachineryKmh : kind === "person" ? maxPersonKmh : null;
    if (limit != null && kmh > limit) {
      if (level !== "alert") level = "warn";
      issues.push(`Exceso velocidad #${tid}: ${kmh} km/h (límite ${limit})`);
    }
  }

  const knEvents = near.filter((e) => e.type === "knowledge_match" || e.type === "knowledge_conjecture");
  if (knEvents.length && level === "ok") level = "warn";

  let headline;
  if (issues.length) headline = issues[0];
  else if (persons || vehicles) headline = `Escena estable — ${persons} persona(s), ${vehicles} máquina(s)`;
  else headline = "Sin personas ni maquinaria detectadas en este instante";

  const glanceParts = [frameRec.time_label || ""];
  if (prox.length) glanceParts.push(`dist. ${prox[0].distance_m} m`);
  else if (persons || vehicles) glanceParts.push(`${persons}👤 · ${vehicles}🚛`);
  const glanceMetric = glanceParts.filter(Boolean).join(" · ");

  const detItems = (frameRec.detections || []).map((det) => ({
    label: det.label || det.kind || "objeto",
    value: `${Math.round((det.confidence || 0) * 100)}%`,
  }));

  const sections = [
    {
      title: "Detecciones",
      items: detItems.length ? detItems : [{ label: "—", value: "Sin detecciones" }],
    },
    {
      title: "Cinemática",
      items: speedLines.length
        ? speedLines.map((s) => ({ label: `#${s.tid}`, value: `${s.kmh} km/h (${s.kind})` }))
        : [{ label: "—", value: "Sin velocidad medible" }],
    },
    {
      title: "Proximidad persona–maquinaria",
      items: prox.length
        ? prox.slice(0, 4).map((p) => ({
            label: `#${p.person_track} ↔ #${p.machinery_track}`,
            value: `${p.distance_m} m`,
            severity: p.alert ? "alert" : null,
          }))
        : [{ label: "—", value: "Sin pares persona–máquina" }],
    },
  ];

  if (near.length) {
    sections.push({
      title: "Eventos en este instante",
      items: near.slice(0, 6).map((ev) => ({
        label: (ev.type || "evento").replace(/_/g, " "),
        value: ev.message || "—",
        severity: ev.severity,
      })),
    });
  }

  const knItems = [];
  for (const ev of knEvents.slice(0, 3)) {
    knItems.push({
      label: (ev.type || "").replace(/_/g, " "),
      value: ev.message || "—",
      severity: ev.type === "knowledge_conjecture" ? "warn" : "ok",
    });
  }
  for (const m of knowledgeMatches.slice(0, 2)) {
    knItems.push({
      label: m.conjecture ? "Conjetura" : "Coincidencia",
      value: `${m.title} (${m.confidence_pct}%)`,
      severity: m.conjecture ? "warn" : "ok",
    });
  }
  if (knItems.length) sections.push({ title: "Biblioteca de aprendizaje", items: knItems });

  if (videoCaption?.caption) {
    sections.unshift({
      title: "IA visual — este instante",
      items: [{ label: videoCaption.time_label || "video", value: videoCaption.caption }],
    });
  }

  const statusMap = { alert: "Riesgo alto", warn: "Atención", ok: "Estable", idle: "Sin datos" };
  return {
    level,
    statusLabel: statusMap[level] || "—",
    headline,
    glanceMetric,
    sections,
    eventsNear: near,
  };
}

export function nearestVideoCaption(videoAi, timeSec, tolerance = 2) {
  const captions = videoAi?.captions || [];
  if (!captions.length) return null;
  let best = captions[0];
  let bestDt = Math.abs((best.time_sec || 0) - timeSec);
  for (const cap of captions) {
    const dt = Math.abs((cap.time_sec || 0) - timeSec);
    if (dt < bestDt) {
      bestDt = dt;
      best = cap;
    }
  }
  return bestDt <= tolerance ? best : null;
}

export function clinicalProgressMessage(raw, progress = null) {
  const text = (raw || "").trim();
  if (!text) return "Preparando auditoría forense…";
  const low = text.toLowerCase();
  if (text === "En cola") return "En cola — preparando revisión forense del video";
  if (text === "Preparando fuentes de video") return "Preparando fuentes de video para auditoría";
  if (text.startsWith("Consultando biblioteca")) return "Contrastando escena con biblioteca de incidentes";
  if (text.startsWith("Analizando video con IA visual")) return "La IA revisa fotogramas clave del video";
  if (text.startsWith("Generando informes")) return "Redactando informe clínico del caso";
  if (text === "Completado") return "Auditoría completa — informe listo para revisión";
  if (text === "Error") return "La auditoría no pudo completarse";
  if (low.includes("calculando cinemática") || low.includes("mapa de calor")) {
    return "Cuantificando velocidades y zonas de tránsito";
  }
  if (low.includes("frame ") && text.includes("/")) {
    return text.replace(/frame /gi, "fotograma ").replace(/Cám\./g, "cámara");
  }
  if (progress != null && progress < 15) return `Iniciando lectura del video — ${text}`;
  return text;
}

export function renderClinicalAuditPanel(container, audit) {
  if (!container) return;
  container.innerHTML = "";
  for (const section of audit.sections || []) {
    const sec = document.createElement("div");
    sec.className = "clinical-section";
    const h = document.createElement("h4");
    h.textContent = section.title;
    sec.appendChild(h);
    const ul = document.createElement("ul");
    ul.className = "clinical-items";
    for (const item of section.items || []) {
      const li = document.createElement("li");
      if (item.severity) li.classList.add(`sev-${item.severity}`);
      li.innerHTML = `<span class="clinical-item-label">${item.label}</span><span class="clinical-item-value">${item.value}</span>`;
      ul.appendChild(li);
    }
    sec.appendChild(ul);
    container.appendChild(sec);
  }
  if (!audit.sections?.length) {
    container.textContent = "Sin datos de auditoría para este instante.";
  }
}

File: forense/app/clinical_eye.py
"""Evaluación clínica del instante forense (ojo clínico)."""

from __future__ import annotations

from typing import Any


def _track_kind_map(frame_rec: dict[str, Any]) -> dict[int, str]:
    out: dict[int, str] = {}
    for tr in frame_rec.get("tracks") or []:
        tid = tr.get("track_id")
        if tid is not None:
            out[int(tid)] = str(tr.get("kind") or "other")
    return out


def _events_near(
    timeline: list[dict[str, Any]] | None,
    time_sec: float,
    *,
    tolerance: float = 1.5,
) -> list[dict[str, Any]]:
    if not timeline:
        return []
    out: list[dict[str, Any]] = []
    for ev in timeline:
        t = float(ev.get("time_sec") or 0)
        if abs(t - time_sec) <= tolerance:
            out.append(ev)
    return out


def evaluate_instant_audit(
    frame_rec: dict[str, Any] | None,
    *,
    timeline: list[dict[str, Any]] | None = None,
    knowledge_matches: list[dict[str, Any]] | None = None,
    video_caption: dict[str, Any] | None = None,
    time_sec: float | None = None,
    min_distance_m: float = 2.0,
    max_machinery_kmh: float = 15.0,
    max_person_kmh: float = 8.0,
    time_tolerance: float = 1.5,
) -> dict[str, Any]:
    """Resume el estado clínico de un fotograma para UI de auditoría."""
    if not frame_rec:
        return {
            "level": "idle",
            "status_label": "Sin datos",
            "headline": "Esperando análisis del video…",
            "glance_metric": "",
            "sections": [],
            "events_near": [],
        }

    t = float(time_sec if time_sec is not None else frame_rec.get("time_sec") or 0)
    counts = frame_rec.get("counts") or {}
    persons = int(counts.get("persons") or 0)
    vehicles = int(counts.get("vehicles") or 0)
    prox = list(frame_rec.get("proximity") or [])
    speeds = list(frame_rec.get("speeds") or [])
    kind_by_track = _track_kind_map(frame_rec)
    events_near = _events_near(timeline, t, tolerance=time_tolerance)

    level = "ok"
    issues: list[str] = []

    prox_alert = next((p for p in prox if p.get("alert")), None)
    if prox_alert:
        level = "alert"
        issues.append(
            f"Proximidad crítica {prox_alert.get('distance_m')} m "
            f"(persona #{prox_alert.get('person_track')} – máq. #{prox_alert.get('machinery_track')})"
        )
    elif prox:
        closest = prox[0]
        dist = float(closest.get("distance_m") or 99)
        if dist <= min_distance_m * 1.5:
            level = "warn"
            issues.append(f"Distancia persona–maquinaria: {dist} m (umbral {min_distance_m} m)")

    speed_lines: list[str] = []
    for sp in speeds:
        tid = int(sp.get("track_id") or 0)
        kmh = float(sp.get("speed_kmh") or 0)
        kind = kind_by_track.get(tid, "other")
        speed_lines.append(f"#{tid} {kmh} km/h ({kind})")
        limit = max_machinery_kmh if kind == "machinery" else max_person_kmh if kind == "person" else None
        if limit is not None and kmh > limit:
            if level != "alert":
                level = "warn"
            issues.append(f"Exceso velocidad #{tid}: {kmh} km/h (límite {limit})")

    kn_events = [e for e in events_near if e.get("type") in ("knowledge_match", "knowledge_conjecture")]
    if kn_events and level == "ok":
        level = "warn"

    if issues:
        headline = issues[0]
    elif persons or vehicles:
        headline = f"Escena estable — {persons} persona(s), {vehicles} máquina(s)"
    else:
        headline = "Sin personas ni maquinaria detectadas en este instante"

    glance_parts = [frame_rec.get("time_label") or ""]
    if prox:
        glance_parts.append(f"dist. {prox[0].get('distance_m')} m")
    elif persons or vehicles:
        glance_parts.append(f"{persons}👤 · {vehicles}🚛")
    glance_metric = " · ".join(p for p in glance_parts if p)

    det_items = []
    for det in frame_rec.get("detections") or []:
        conf = int(round(float(det.get("confidence") or 0) * 100))
        det_items.append(
            {
                "label": det.get("label") or det.get("kind") or "objeto",
                "value": f"{conf}%",
            }
        )

    sections: list[dict[str, Any]] = [
        {
            "title": "Detecciones",
            "items": det_items or [{"label": "—", "value": "Sin detecciones"}],
        },
        {
            "title": "Cinemática",
            "items": (
                [{"label": line.split()[0], "value": " ".join(line.split()[1:])} for line in speed_lines]
                if speed_lines
                else [{"label": "—", "value": "Sin velocidad medible"}]
            ),
        },
        {
            "title": "Proximidad persona–maquinaria",
            "items": (
                [
                    {
                        "label": f"#{p.get('person_track')} ↔ #{p.get('machinery_track')}",
                        "value": f"{p.get('distance_m')} m",
                        "severity": "alert" if p.get("alert") else None,
                    }
                    for p in prox[:4]
                ]
                if prox
                else [{"label": "—", "value": "Sin pares persona–máquina"}]
            ),
        },
    ]

    if events_near:
        sections.append(
            {
                "title": "Eventos en este instante",
                "items": [
                    {
                        "label": ev.get("type", "evento").replace("_", " "),
                        "value": ev.get("message") or "—",
                        "severity": ev.get("severity"),
                    }
                    for ev in events_near[:6]
                ],
            }
        )

    kn_items: list[dict[str, Any]] = []
    for ev in kn_events[:3]:
        kn_items.append(
            {
                "label": ev.get("type", "").replace("_", " "),
                "value": ev.get("message") or "—",
                "severity": "warn" if ev.get("type") == "knowledge_conjecture" else "ok",
            }
        )
    for m in (knowledge_matches or [])[:2]:
        tag = "Conjetura" if m.get("conjecture") else "Coincidencia"
        kn_items.append(
            {
                "label": tag,
                "value": f"{m.get('title')} ({m.get('confidence_pct')}%)",
                "severity": "warn" if m.get("conjecture") else "ok",
            }
        )
    if kn_items:
        sections.append({"title": "Biblioteca de aprendizaje", "items": kn_items})

    if video_caption and (video_caption.get("caption") or "").strip():
        sections.insert(
            0,
            {
                "title": "IA visual — este instante",
                "items": [
                    {
                        "label": video_caption.get("time_label") or "video",
                        "value": video_caption.get("caption"),
                        "severity": None,
                    }
                ],
            },
        )

    status_map = {
        "alert": "Riesgo alto",
        "warn": "Atención",
        "ok": "Estable",
        "idle": "Sin datos",
    }
    return {
        "level": level,
        "status_label": status_map.get(level, "—"),
        "headline": headline,
        "glance_metric": glance_metric,
        "sections": sections,
        "events_near": events_near,
    }


def clinical_progress_message(raw: str, *, progress: int | None = None) -> str:
    """Traduce mensajes técnicos de progreso a copy clínico es-CL."""
    text = (raw or "").strip()
    if not text:
        return "Preparando auditoría forense…"
    low = text.lower()
    if text == "En cola":
        return "En cola — preparando revisión forense del video"
    if text == "Preparando fuentes de video":
        return "Preparando fuentes de video para auditoría"
    if text.startswith("Consultando biblioteca"):
        return "Contrastando escena con biblioteca de incidentes"
    if text.startswith("Analizando video con IA visual"):
        return "La IA revisa fotogramas clave del video"
    if text.startswith("Generando informes"):
        return "Redactando informe clínico del caso"
    if text == "Completado":
        return "Auditoría completa — informe listo para revisión"
    if text == "Error":
        return "La auditoría no pudo completarse"
    if "calculando cinemática" in low or "mapa de calor" in low:
        return "Cuantificando velocidades y zonas de tránsito"
    if "frame " in low and "/" in text:
        return text.replace("frame ", "fotograma ").replace("Cám.", "cámara")
    if progress is not None and progress < 15:
        return f"Iniciando lectura del video — {text}"
    return text
